substrate/runtime/contract/src/exec.rs

```diff
@@ -0,0 +1,257 @@
+					&dest_code,
+					&mut CallContext {
+						ctx: &mut nested,
+						_caller: caller,
+					},
+					gas_meter,
+				).map_err(|_| "vm execute returned error while call")?
+			} else {
+				// that was a plain transfer
+				vm::ExecutionResult {
+					return_data: Vec::new(),
+				}
+			};
+
+			(exec_result, nested.overlay.into_change_set())
+		};
+
+		self.overlay.commit(change_set);
+
+		Ok(CallReceipt {
+			return_data: exec_result.return_data,
+		})
+	}
+
+	pub fn create(
+		&mut self,
+		caller: T::AccountId,
+		endowment: T::Balance,
+		gas_meter: &mut GasMeter<T>,
+		ctor: &[u8],
+		_data: &[u8],
+	) -> Result<CreateReceipt<T>, &'static str> {
+		let create_base_fee = <Module<T>>::create_base_fee();
+		if gas_meter.charge(create_base_fee).is_out_of_gas() {
+			return Err("not enough gas to pay base create fee");
+		}
+
+		let dest = T::DetermineContractAddress::contract_address_for(ctor, &self.self_account);
+		if <CodeOf<T>>::exists(&dest) {
+			// TODO: Is it enough?
+			return Err("contract already exists");
+		}
+
+		let change_set = {
+			let mut overlay = OverlayAccountDb::new(&self.overlay);
+
+			if endowment > T::Balance::zero() {
+				transfer(
+					gas_meter,
+					true,
+					&self.self_account,
+					&dest,
+					endowment,
+					&mut overlay,
+				)?;
+			}
+
+			let mut nested = ExecutionContext {
+				overlay: overlay,
+				self_account: dest.clone(),
+				depth: self.depth + 1,
+			};
+			let exec_result = {
+				vm::execute(
+					ctor,
+					&mut CallContext {
+						ctx: &mut nested,
+						_caller: caller,
+					},
+					gas_meter,
+				).map_err(|_| "vm execute returned error while create")?
+			};
+
+			nested.overlay.set_code(&dest, exec_result.return_data().to_vec());
+			nested.overlay.into_change_set()
+		};
+
+		self.overlay.commit(change_set);
+
+		Ok(CreateReceipt {
+			address: dest,
+		})
+	}
+}
+
+fn transfer<T: Trait>(
+	gas_meter: &mut GasMeter<T>,
+	contract_create: bool,
+	transactor: &T::AccountId,
+	dest: &T::AccountId,
+	value: T::Balance,
+	overlay: &mut OverlayAccountDb<T>,
+) -> Result<(), &'static str> {
+	let would_create = overlay.get_balance(transactor).is_zero();
+
+	let fee: T::Balance = if contract_create {
+		<Module<T>>::contract_fee()
+	} else {
+		if would_create {
+			<staking::Module<T>>::creation_fee()
+		} else {
+			<staking::Module<T>>::transfer_fee()
+		}
+	};
+
+	if gas_meter.charge_by_balance(fee).is_out_of_gas() {
+		return Err("not enough gas to pay transfer fee");
+	}
+
+	let from_balance = overlay.get_balance(transactor);
+	if from_balance < value {
+		return Err("balance too low to send value");
+	}
+	if would_create && value < <staking::Module<T>>::existential_deposit() {
+		return Err("value too low to create account");
+	}
+	if <staking::Module<T>>::bondage(transactor) > <system::Module<T>>::block_number() {
+		return Err("bondage too high to send value");
+	}
+
+	let to_balance = overlay.get_balance(dest);
+	if to_balance + value <= to_balance {
```

Comment:
Just a possible confusion here: it looks like we define `staking::Trait::Balance` to be wrapping (non-panic), but we're using `std` traits `Add/Sub` etc, which by default is non-wrapping (panic). So if someone accidentally set `staking::Trait::Balance` to be, for example, `usize`. The code would panic.

Maybe it's worth considering to define another trait (like `WrappingAdd`?).

Reply:
You're right! Will use `checked_add` here and in the `staking` module.